include/swift/AST/DiagnosticEngine.h

```diff
@@ -100,111 +80,61 @@ namespace swift {
-      const DeclAttribute *DeclAttributeVal;
-      llvm::VersionTuple VersionVal;
-      LayoutConstraint LayoutConstraintVal;
-    };
-    
+    using UnionType =
+        TaggedUnion<void, int, unsigned, StringRef, DeclNameRef, ObjCSelector,
+                    ValueDecl *, Type, TypeRepr *, PatternKind, SelfAccessKind,
+                    ReferenceOwnership, StaticSpellingKind, DescriptiveDeclKind,
+                    const DeclAttribute *, llvm::VersionTuple,
+                    LayoutConstraint>;
+    UnionType Union;
+
   public:
-    DiagnosticArgument(StringRef S)
-      : Kind(DiagnosticArgumentKind::String), StringVal(S) {
-    }
+    DiagnosticArgument(StringRef S) : Union(S) {}
 
-    DiagnosticArgument(int I) 
-      : Kind(DiagnosticArgumentKind::Integer), IntegerVal(I) {
-    }
+    DiagnosticArgument(int I) : Union(I) {}
 
-    DiagnosticArgument(unsigned I) 
-      : Kind(DiagnosticArgumentKind::Unsigned), UnsignedVal(I) {
-    }
+    DiagnosticArgument(unsigned I) : Union(I) {}
 
-    DiagnosticArgument(DeclNameRef R)
-        : Kind(DiagnosticArgumentKind::Identifier), IdentifierVal(R) {}
+    DiagnosticArgument(DeclNameRef R) : Union(R) {}
 
-    DiagnosticArgument(DeclName D)
-        : Kind(DiagnosticArgumentKind::Identifier),
-          IdentifierVal(DeclNameRef(D)) {}
+    DiagnosticArgument(DeclName D) : Union(DeclNameRef(D)) {}
 
-    DiagnosticArgument(DeclBaseName D)
-        : Kind(DiagnosticArgumentKind::Identifier),
-          IdentifierVal(DeclNameRef(D)) {}
+    DiagnosticArgument(DeclBaseName D) : Union(DeclNameRef(D)) {}
 
-    DiagnosticArgument(Identifier I)
-      : Kind(DiagnosticArgumentKind::Identifier),
-        IdentifierVal(DeclNameRef(I)) {
-    }
+    DiagnosticArgument(Identifier I) : Union(DeclNameRef(I)) {}
 
-    DiagnosticArgument(ObjCSelector S)
-      : Kind(DiagnosticArgumentKind::ObjCSelector), ObjCSelectorVal(S) {
-    }
+    DiagnosticArgument(ObjCSelector S) : Union(S) {}
 
-    DiagnosticArgument(ValueDecl *VD)
-      : Kind(DiagnosticArgumentKind::ValueDecl), TheValueDecl(VD) {
-    }
+    DiagnosticArgument(ValueDecl *VD) : Union(VD) {}
 
-    DiagnosticArgument(Type T)
-      : Kind(DiagnosticArgumentKind::Type), TypeVal(T) {
-    }
+    DiagnosticArgument(Type T) : Union(T) {}
 
-    DiagnosticArgument(TypeRepr *T)
-      : Kind(DiagnosticArgumentKind::TypeRepr), TyR(T) {
-    }
+    DiagnosticArgument(TypeRepr *T) : Union(T) {}
 
     DiagnosticArgument(const TypeLoc &TL) {
       if (TypeRepr *tyR = TL.getTypeRepr()) {
-        Kind = DiagnosticArgumentKind::TypeRepr;
-        TyR = tyR;
+        Union = UnionType(tyR);
       } else {
-        Kind = DiagnosticArgumentKind::Type;
-        TypeVal = TL.getType();
+        Union = UnionType(TL.getType());
       }
     }
 
-    DiagnosticArgument(PatternKind K)
-        : Kind(DiagnosticArgumentKind::PatternKind), PatternKindVal(K) {}
+    DiagnosticArgument(PatternKind K) : Union(K) {}
```

Comment:
Unfortunately we need to keep a bunch of these initializers around for overload resolution